srai/regionizers/h3_regionizer.py

```diff
@@ -0,0 +1,191 @@
+        and optionally applies buffering.
+
+        Args:
+            gdf (gpd.GeoDataFrame): (Multi)Polygons to be regionized.
+
+        Returns:
+            gpd.GeoDataFrame: H3 cells.
+
+        Raises:
+            ValueError: If provided GeoDataFrame has no crs defined.
+
+        """
+        gdf_wgs84 = gdf.to_crs(epsg=4326)
+        # transform multipolygons to multiple polygons
+        gdf_exploded = gdf_wgs84.explode(index_parts=True).reset_index(drop=True)
+
+        h3_indexes = (
+            seq(gdf_exploded["geometry"])
+            .map(self._polygon_shapely_to_h3)
+            .flat_map(lambda polygon: h3.polygon_to_cells(polygon, self.resolution))
+            .to_list()
+        )
+
+        gdf_h3 = self._gdf_from_h3_indexes(h3_indexes)
+        if self.buffer:
+            return self._buffer(gdf_exploded, gdf_h3)
+
+        return gdf_h3.to_crs(gdf.crs)
+
+    def _gdf_from_h3_indexes(self, h3_indexes: List[str]) -> gpd.GeoDataFrame:
+        """
+        Convert H3 Indexes to GeoDataFrame with geometries.
+
+        Args:
+            h3_indexes (List[str]): H3 Indexes.
+
+        Returns:
+            gpd.GeoDataFrame: H3 cells.
+
+        """
+        return gpd.GeoDataFrame(
+            None,
+            index=h3_indexes,
+            geometry=[self._h3_index_to_shapely_polygon(h3_index) for h3_index in h3_indexes],
+            crs="epsg:4326",
+        )
+
+    def _h3_index_to_shapely_polygon(self, h3_index: str) -> geometry.Polygon:
+        """
+        Convert H3 Index to Shapely polygon.
+
+        Args:
+            h3_index (str): H3 Index to be converted.
+
+        Returns:
+            geometry.Polygon: Converted polygon.
+
+        """
+        return self._polygon_h3_to_shapely(h3.cells_to_polygons([h3_index])[0])
+
+    def _polygon_h3_to_shapely(self, polygon: h3.Polygon) -> geometry.Polygon:
+        """
+        Convert H3 Polygon to Shapely Polygon.
+
+        Args:
+            polygon (h3.Polygon): H3 Polygon to be converted.
+
+        Returns:
+            geometry.Polygon: Converted polygon.
+
+        """
+        return geometry.Polygon(
+            shell=[coord[::-1] for coord in polygon.outer],
+            holes=[[coord[::-1] for coord in hole] for hole in polygon.holes],
+        )
+
+    def _polygon_shapely_to_h3(self, polygon: geometry.Polygon) -> h3.Polygon:
+        """
+        Convert Shapely Polygon to H3 Polygon.
+
+        Args:
+            polygon (geometry.Polygon): Shapely polygon to be converted.
+
+        Returns:
+            h3.Polygon: Converted polygon.
+
+        """
+        exterior = [coord[::-1] for coord in list(polygon.exterior.coords)]
+        interiors = [
+            [coord[::-1] for coord in list(interior.coords)] for interior in polygon.interiors
+        ]
+        return h3.Polygon(exterior, *interiors)
+
+    def _buffer(self, gdf: gpd.GeoDataFrame, gdf_h3: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
```

Comment:
Wouldn't it be faster to buffer the original geometries with `shapely` by an amount equal to half the size of a hex on a given resolution (not sure if half or full radius)?
@simonusher was doing this in his original code in a concise way.

Reply:
Yes, it would be a lot faster. I have this implementation myself too, but it is unfortunately not ideal and does not always give correct results. The problem is that the hexagons' edge lengths on the same resolution differ significantly, and the average may be well off compared to the factual edge lengths in obtained hexagons. 
But as I said, I need to work on that a bit more and maybe I will fall back to this solution.

Reply:
This may be a silly question, but would it work and be relatively fast to just buffer the original geometry with a big margin (2x hex radius?) and fill it with h3 cells, then just make a spatial join with an original geometry? What am I missing here?

Reply:
Yes, I think the big margin would actually solve that. I will try to implement that then

Reply:
Yes, that was my solution to this and I thought we could use that here